File: packages/host4_flutter_aivoice/ios/AiVoice_副本/TokenGenerator/AccessToken.m
//
//  AccessToken.m
//  火山引擎 RTC Token Generator
//  Token 生成和解析实现
//

#import "AccessToken.h"
#import "ByteBuf.h"
#import "RtcUtilTools.h"

@implementation AccessToken

- (instancetype)initWithAppID:(NSString *)appID
                       appKey:(NSString *)appKey
                       roomID:(NSString *)roomID
                       userID:(NSString *)userID {
    self = [super init];
    if (self) {
        _appID = appID;
        _appKey = appKey;
        _roomID = roomID;
        _userID = userID;
        _issuedAt = [RtcUtilTools getTimestamp];
        _nonce = [RtcUtilTools randomInt];
        _privileges = [NSMutableDictionary dictionary];
        _expireAt = 0;
    }
    return self;
}

+(NSString *)generateToken:(NSString *)roomID userID:(NSString *)userID{
    
    AccessToken *token = [[AccessToken alloc] initWithAppID:AIVoiceAppId appKey:AIVoiceAppKey roomID:roomID userID:userID];
    
    [token setExpireTime:[RtcUtilTools getTimestamp] + 3600];
    [token addPrivilege:RTCPrivilegeSubscribeStream expireTimestamp:0];
    [token addPrivilege:RTCPrivilegePublishStream expireTimestamp:[RtcUtilTools getTimestamp] + 3600];
    
    return [token serialize];
}

+ (NSString *)getVersion {
    return @"001";
}

- (void)addPrivilege:(RTCPrivilege)privilege expireTimestamp:(int32_t)expireTimestamp {
    self.privileges[@(privilege)] = @(expireTimestamp);
    
    // 如果添加发布流权限，自动添加相关子权限
    if (privilege == RTCPrivilegePublishStream) {
        self.privileges[@(RTCPrivilegePublishAudioStream)] = @(expireTimestamp);
        self.privileges[@(RTCPrivilegePublishVideoStream)] = @(expireTimestamp);
        self.privileges[@(RTCPrivilegePublishDataStream)] = @(expireTimestamp);
    }
}

- (void)setExpireTime:(int32_t)expireTimestamp {
    self.expireAt = expireTimestamp;
}

- (NSData *)packMsg {
    ByteBuf *buffer = [[ByteBuf alloc] init];
    
    [buffer putInt:self.nonce];
    [buffer putInt:self.issuedAt];
    [buffer putInt:self.expireAt];
    [buffer putString:self.roomID];
    [buffer putString:self.userID];
    [buffer putIntMap:self.privileges];
    
    return [buffer asBytes];
}

- (NSString *)serialize {
    NSData *msg = [self packMsg];
    
    // 生成签名
    self.signature = [RtcUtilTools hmacSign:self.appKey data:msg];
    
    if (!self.signature) {
        NSLog(@"❌ AccessToken: 签名生成失败");
        return nil;
    }
    
    // 打包消息和签名
    ByteBuf *buffer = [[ByteBuf alloc] init];
    [buffer putBytes:msg];
    [buffer putBytes:self.signature];
    
    NSData *content = [buffer asBytes];
    
    // 生成最终 Token: 版本号 + AppID + Base64(内容)
    NSString *encoded = [RtcUtilTools base64Encode:content];
    
    return [NSString stringWithFormat:@"%@%@%@", 
            [AccessToken getVersion], 
            self.appID, 
            encoded];
}

- (BOOL)verify:(NSString *)key {
    // 检查是否过期
    if (self.expireAt > 0 && [RtcUtilTools getTimestamp] > self.expireAt) {
        return NO;
    }
    
    self.appKey = key;
    
    // 重新生成签名并比较
    NSData *newSignature = [RtcUtilTools hmacSign:self.appKey data:[self packMsg]];
    
    if (!newSignature || !self.signature) {
        return NO;
    }
    
    return [self.signature isEqualToData:newSignature];
}

+ (AccessToken *)parse:(NSString *)tokenString {
    AccessToken *token = [[AccessToken alloc] initWithAppID:@"" 
                                                     appKey:@"" 
                                                     roomID:@"" 
                                                     userID:@""];
    
    if (!tokenString || tokenString.length <= RTCVersionLength + RTCAppIDLength) {
        return token;
    }
    
    // 验证版本号
    NSString *version = [tokenString substringToIndex:RTCVersionLength];
    if (![version isEqualToString:[AccessToken getVersion]]) {
        return token;
    }
    
    // 提取 AppID
    NSRange appIDRange = NSMakeRange(RTCVersionLength, RTCAppIDLength);
    token.appID = [tokenString substringWithRange:appIDRange];
    
    // 提取并解码内容
    NSString *encodedContent = [tokenString substringFromIndex:RTCVersionLength + RTCAppIDLength];
    NSData *content = [RtcUtilTools base64Decode:encodedContent];
    
    if (!content) {
        return token;
    }
    
    // 解析内容
    ByteBuf *buffer = [[ByteBuf alloc] initWithBytes:content];
    
    NSData *msg = [buffer readBytes];
    token.signature = [buffer readBytes];
    
    // 解析消息
    ByteBuf *msgBuf = [[ByteBuf alloc] initWithBytes:msg];
    
    token.nonce = [msgBuf readInt];
    token.issuedAt = [msgBuf readInt];
    token.expireAt = [msgBuf readInt];
    token.roomID = [msgBuf readString];
    token.userID = [msgBuf readString];
    token.privileges = [[msgBuf readIntMap] mutableCopy];
    
    return token;
}

- (NSString *)description {
    return [NSString stringWithFormat:@"AccessToken{\n"
            "  appID='%@',\n"
            "  roomID='%@',\n"
            "  userID='%@',\n"
            "  issuedAt=%d,\n"
            "  expireAt=%d,\n"
            "  nonce=%d,\n"
            "  privileges=%@\n"
            "}",
            self.appID,
            self.roomID,
            self.userID,
            self.issuedAt,
            self.expireAt,
            self.nonce,
            self.privileges];
}

@end

File: packages/host4_flutter_aivoice/ios/Classes/Utils/AccessToken.swift
import Foundation

// MARK: - RTC 权限枚举
enum RTCPrivilege: Int16 {
  case publishStream      = 0  /// 发布流
  case publishAudioStream = 1  /// 发布音频流（内部）
  case publishVideoStream = 2  /// 发布视频流（内部）
  case publishDataStream  = 3  /// 发布数据流（内部）
  case subscribeStream    = 4  /// 订阅流
}

/// 火山引擎 RTC Token 生成器
class AccessToken {
  static let version = "001"
  static let versionLength = 3
  static let appIDLength = 24

  var appID: String
  var appKey: String
  var roomID: String
  var userID: String
  var issuedAt: Int32
  var expireAt: Int32
  var nonce: Int32
  var privileges: [Int: Int] = [:]
  var signature: Data?

  init(appID: String, appKey: String, roomID: String, userID: String) {
    self.appID = appID
    self.appKey = appKey
    self.roomID = roomID
    self.userID = userID
    issuedAt = Int32(RtcUtils.timestamp())
    nonce = Int32(RtcUtils.randomInt())
    expireAt = 0
  }

  /// 生成 RTC Token（自包含 appId + appKey）
  static func generate(roomID: String, userID: String) -> String? {
    let appID = "68230496df1dcd01804db0a9"
    let appKey = "5054eb5367ec4703bc6763ca78736038"

    let token = AccessToken(appID: appID, appKey: appKey,
                            roomID: roomID, userID: userID)
    token.setExpire(time: Int32(RtcUtils.timestamp()) + 3600)
    token.addPrivilege(.subscribeStream, expire: 0)
    token.addPrivilege(.publishStream, expire: Int32(RtcUtils.timestamp()) + 3600)

    return token.serialize()
  }

  func addPrivilege(_ privilege: RTCPrivilege, expire: Int32) {
    privileges[Int(privilege.rawValue)] = Int(expire)
    if privilege == .publishStream {
      privileges[Int(RTCPrivilege.publishAudioStream.rawValue)] = Int(expire)
      privileges[Int(RTCPrivilege.publishVideoStream.rawValue)] = Int(expire)
      privileges[Int(RTCPrivilege.publishDataStream.rawValue)] = Int(expire)
    }
  }

  func setExpire(time: Int32) {
    expireAt = time
  }

  func packMsg() -> Data {
    let buf = ByteBuf()
    buf.putInt(nonce)
         .putInt(issuedAt)
         .putInt(expireAt)
         .putString(roomID)
         .putString(userID)
         .putIntMap(privileges)
    return buf.bytes
  }

  func serialize() -> String? {
    let msg = packMsg()
    signature = RtcUtils.hmacSha256(key: appKey, message: msg)

    let buf = ByteBuf()
    buf.putBytes(msg)
    if let sig = signature { buf.putBytes(sig) }

    let encoded = RtcUtils.base64Encode(buf.bytes)
    return "\(Self.version)\(appID)\(encoded)"
  }

  func verify(key: String) -> Bool {
    if expireAt > 0, Int32(RtcUtils.timestamp()) > expireAt { return false }
    appKey = key
    guard let sig = signature else { return false }
    let newSig = RtcUtils.hmacSha256(key: appKey, message: packMsg())
    return sig == newSig
  }

  static func parse(from tokenString: String) -> AccessToken {
    let token = AccessToken(appID: "", appKey: "", roomID: "", userID: "")
    guard tokenString.count > versionLength + appIDLength else { return token }

    let ver = String(tokenString.prefix(versionLength))
    guard ver == version else { return token }

    let appIDStart = tokenString.index(tokenString.startIndex, offsetBy: versionLength)
    let appIDEnd = tokenString.index(appIDStart, offsetBy: appIDLength)
    token.appID = String(tokenString[appIDStart..<appIDEnd])

    let encoded = String(tokenString[appIDEnd...])
    guard let content = RtcUtils.base64Decode(encoded) else { return token }

    let buf = ByteBuf(data: content)
    let msg = buf.readBytes()
    token.signature = buf.readBytes()

    let msgBuf = ByteBuf(data: msg)
    token.nonce = msgBuf.readInt()
    token.issuedAt = msgBuf.readInt()
    token.expireAt = msgBuf.readInt()
    token.roomID = msgBuf.readString()
    token.userID = msgBuf.readString()
    token.privileges = msgBuf.readIntMap()

    return token
  }
}

extension AccessToken: CustomStringConvertible {
  var description: String {
    """
    AccessToken{
      appID='\(appID)',
      roomID='\(roomID)',
      userID='\(userID)',
      issuedAt=\(issuedAt),
      expireAt=\(expireAt),
      nonce=\(nonce),
      privileges=\(privileges)
    }
    """
  }
}
